File: tests/LevenbergMarquardtTests.cpp
#include <gtest/gtest.h>

#include "clf/DenseLevenbergMarquardt.hpp"
#include "clf/SparseLevenbergMarquardt.hpp"

#include "TestPenaltyFunctions.hpp"

using namespace clf;

TEST(LevenbergMarquardtTests, DenseMatrices) {
  auto para = std::make_shared<Parameters>();

  // create the example penalty functions
  auto func0 = std::make_shared<tests::DensePenaltyFunctionTest0>();
  auto func1 = std::make_shared<tests::DensePenaltyFunctionTest1>();
  auto cost = std::make_shared<DenseCostFunction>(DensePenaltyFunctions({func0, func1}));

  DenseLevenbergMarquardt lm(cost, para);
  EXPECT_EQ(lm.NumParameters(), cost->InputDimension());

  Eigen::VectorXd beta = Eigen::VectorXd::Random(lm.NumParameters());
  Eigen::VectorXd costVec;
  const std::pair<Optimization::Convergence, double> result = lm.Minimize(beta, costVec);
  EXPECT_TRUE(result.first>0);

  Eigen::VectorXd expected(3);
  expected << 0.0, 1.0, 0.0;
  EXPECT_NEAR((beta-expected).norm(), 0.0, 1.0e-10);
  EXPECT_NEAR(costVec.norm(), 0.0, 1.0e-10);
  EXPECT_NEAR(cost->Gradient(beta).norm(), 0.0, 1.0e-10);
}

TEST(LevenbergMarquardtTests, SparseMatrices) {
  auto para = std::make_shared<Parameters>();

  // create the example penalty functions
  auto func0 = std::make_shared<tests::SparsePenaltyFunctionTest0>();
  auto func1 = std::make_shared<tests::SparsePenaltyFunctionTest1>();
  auto cost = std::make_shared<SparseCostFunction>(SparsePenaltyFunctions({func0, func1}));

  SparseLevenbergMarquardt lm(cost, para);
  EXPECT_EQ(lm.NumParameters(), cost->InputDimension());

  Eigen::VectorXd beta = Eigen::VectorXd::Random(lm.NumParameters());
  Eigen::VectorXd costVec;
  const std::pair<Optimization::Convergence, double> result = lm.Minimize(beta, costVec);
  EXPECT_TRUE(result.first>0);

  Eigen::VectorXd expected(3);
  expected << 0.0, 1.0, 0.0;
  EXPECT_NEAR((beta-expected).norm(), 0.0, 1.0e-10);
  EXPECT_NEAR(costVec.norm(), 0.0, 1.0e-10);
  EXPECT_NEAR(cost->Gradient(beta).norm(), 0.0, 1.0e-10);
}

File: tests/LevenbergMarquardtTests.py
"""! @brief Test the Levenberg Marquardt algorithm (clf::LevenbergMarquardt) python interface"""

import unittest

import numpy as np
import scipy

import random

from TestPenaltyFunctions import *
import PyCoupledLocalFunctions as clf

class TestLevenbergMarquardt(unittest.TestCase):
    """! Test the Levenberg Marquardt (clf::LevenbergMarquardt) python interface"""

    def test_dense(self):
        para = clf.Parameters()

        func0 = DensePenaltyFunctionTest0()
        func1 = DensePenaltyFunctionTest1()
        cost = clf.DenseCostFunction([func0, func1])

        lm = clf.DenseLevenbergMarquardt(cost, para)
        self.assertEqual(lm.NumParameters(), cost.InputDimension())

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(lm.NumParameters())])
        result = lm.Minimize(beta)
        self.assertEqual(len(result), 4)
        self.assertTrue(result[0]==clf.OptimizationConvergence.CONVERGED or
                        result[0]==clf.OptimizationConvergence.CONVERGED_FUNCTION_SMALL or
                        result[0]==clf.OptimizationConvergence.CONVERGED_GRADIENT_SMALL)
        self.assertAlmostEqual(result[1], 0.0)
        expected = np.array([0.0, 1.0, 0.0])
        self.assertAlmostEqual(np.linalg.norm(result[2]-expected), 0.0)
        self.assertAlmostEqual(np.linalg.norm(result[3]), 0.0)

    def test_sparse(self):
        para = clf.Parameters()

        func0 = SparsePenaltyFunctionTest0()
        func1 = SparsePenaltyFunctionTest1()
        cost = clf.SparseCostFunction([func0, func1])

        lm = clf.SparseLevenbergMarquardt(cost, para)
        self.assertEqual(lm.NumParameters(), cost.InputDimension())

        beta = np.array([random.uniform(-1.0, 1.0) for i in range(lm.NumParameters())])
